File: python/quick-sort.py
import numpy as np

array = np.random.randint(1, 101, 100)

def quick_sort(array, firstIndex, lastIndex):
    if firstIndex >= lastIndex:
        return
    pivot = array[lastIndex]
    leftPointer = firstIndex
    rightPointer = lastIndex

    while leftPointer < rightPointer:
        while array[leftPointer] <= pivot and leftPointer < rightPointer:
            leftPointer+=1
        while array[rightPointer] >= pivot and leftPointer < rightPointer:
            rightPointer-=1
        swap(array, leftPointer, rightPointer)
    swap(array, leftPointer, lastIndex)

    quick_sort(array, 0, leftPointer - 1)
    quick_sort(array, leftPointer + 1, lastIndex)

def swap(array, a, b):
    tmp = array[a]
    array[a] = array[b]
    array[b] = tmp

print('Array', array)
quick_sort(array, 0, len(array) - 1)
print('Sorted array', array)

File: javascript/quick-sort.js
function quickSort(array, lowIndex, highIndex) {
    if (lowIndex >= highIndex) {
        return;
    }
    const pivot = array[highIndex];
    let lp = lowIndex;
    let rp = highIndex;
    while (lp < rp) {
        while (array[lp] <= pivot && lp < rp) {
            lp++
        }
        while (array[rp] >= pivot && lp < rp) {
            rp--;
        }
        swap(array, lp, rp);
    }
    swap(array, lp, highIndex);

    quickSort(array, lowIndex, lp - 1);
    quickSort(array, lp + 1, highIndex);
}

function swap(array, i, j) {
    const tmp = array[i];
    array[i] = array[j];
    array[j] = tmp;
}